File: SearchA2DMatrix.cpp
#include <iostream>
using namespace std;

class Solution 
{
public:
    bool searchMatrix(vector<vector<int> > &matrix, int target) 
    {
    	if (matrix.size() == 0)
    		return false;
    	int size = matrix.size();
    	int l = 0;
    	int r = size - 1;
    	int m;
    	while (l <= r)
    	{
    		m = (l + r) / 2;
    		if (matrix[m][0] == target)
    			return true;
    		else if (matrix[m][0] > target)
    			r = m - 1;
    		else
    			l = m + 1;
    	}
    	int row = r;
    	if (r < 0)
    		return false;
    	l = 0;
    	r = matrix[0].size() - 1;
    	while (l <= r)
    	{
    		m = (l + r) / 2;
    		if (matrix[row][m] == target)
    			return true;
    		else if (matrix[row][m] > target)
    			r = m - 1;
    		else
    			l = m + 1;
    	}
    	return false;
    }
};

int main()
{
	return 0;
}

File: JavaSolution/src/Searcha2DMatrix.java
/**
 * Created by xyunpeng on 2/3/16.
 */
public class Searcha2DMatrix {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return false;
        }

        int m = matrix[0].length;
        int start = 0;
        int end = matrix.length * m - 1;
        int mid;
        while (start + 1 < end) {
            mid = start + ((end - start) >> 1);
            int t = matrix[mid / m][mid % m];
            if (t == target) {
                return true;
            } else if ( t < target) {
                start = mid;
            } else {
                end = mid;
            }
        }

        if (matrix[start / m][start % m] == target) {
            return true;
        }

        if (matrix[end / m][end % m] == target) {
            return true;
        }

        return false;
    }
}
